# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/fetch/InternalResponse.cpp
# Commit: ffcced3e103d
# Full Hash: ffcced3e103d1563abe464e65934ae2d3593db61
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2023-06-01 09:35:25
# Description:
#   Bug 1833795 - Handle the case that InternalResponse.body/alternativeBody serialization fail. r=dom-worker-reviewers,smaug
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D178788
# ==============================================================================

diff -r 1b938c15a507 -r ffcced3e103d dom/fetch/InternalResponse.cpp
--- a/dom/fetch/InternalResponse.cpp	Thu Jun 01 01:45:36 2023 +0300
+++ b/dom/fetch/InternalResponse.cpp	Wed May 31 23:14:11 2023 +0000
@@ -207,16 +207,30 @@
   GetUnfilteredBody(getter_AddRefs(body), &bodySize);
 
   if (body) {
-    result.body() = Some(ToParentToChildStream(
-        WrapNotNull(body), bodySize, aBackgroundParent, mSerializeAsLazy));
-    result.bodySize() = bodySize;
+    ParentToChildStream bodyStream = ToParentToChildStream(
+        WrapNotNull(body), bodySize, aBackgroundParent, mSerializeAsLazy);
+    // The body stream can fail to serialize as an IPCStream. In the case, the
+    // IPCStream's type would be T__None. Don't set up IPCInternalResponse's
+    // body with the failed IPCStream.
+    if (mSerializeAsLazy || bodyStream.get_IPCStream().stream().type() !=
+                                mozilla::ipc::InputStreamParams::T__None) {
+      result.body() = Some(bodyStream);
+      result.bodySize() = bodySize;
+    }
   }
 
   nsCOMPtr<nsIInputStream> alternativeBody = TakeAlternativeBody();
   if (alternativeBody) {
-    result.alternativeBody() = Some(
+    ParentToChildStream alterBodyStream =
         ToParentToChildStream(WrapNotNull(alternativeBody), UNKNOWN_BODY_SIZE,
-                              aBackgroundParent, mSerializeAsLazy));
+                              aBackgroundParent, mSerializeAsLazy);
+    // The body stream can fail to serialize as an IPCStream. In the case, the
+    // IPCStream's type would be T__None. Don't set up IPCInternalResponse's
+    // body with the failed IPCStream.
+    if (mSerializeAsLazy || alterBodyStream.get_IPCStream().stream().type() !=
+                                mozilla::ipc::InputStreamParams::T__None) {
+      result.alternativeBody() = Some(alterBodyStream);
+    }
   }
 
   return result;